server/websocket.cpp

```diff
@@ -0,0 +1,213 @@
+        ws_.set_option(
+            websocket::stream_base::timeout::suggested(
+                beast::role_type::server));
+
+        // Set a decorator to change the Server of the handshake
+        ws_.set_option(websocket::stream_base::decorator(
+            [](websocket::response_type& res)
+            {
+                res.set(http::field::server,
+                    std::string(BOOST_BEAST_VERSION_STRING) +
+                        " websocket-server-async");
+            }));
+        // Accept the websocket handshake
+        ws_.async_accept(
+            beast::bind_front_handler(
+                &DfWebsocket::OnAccept,
+                shared_from_this()));
+    }
+
+    void DoRead() {
+        // Read a message into our buffer
+        ws_.async_read(
+            buffer_,
+            beast::bind_front_handler(
+                &DfWebsocket::OnRead,
+                shared_from_this()));
+    }
+
+    void OnAccept(beast::error_code ec)
+    {
+        if(ec)
+            return fail(ec, "accept");
+
+        onOpenHandler(socketID_);
+
+        DoRead();
+    }
+
+    void
+    OnRead(
+        beast::error_code ec,
+        std::size_t bytes_transferred)
+    {
+        boost::ignore_unused(bytes_transferred);
+
+        // This indicates that the session was closed
+        if(ec == websocket::error::closed)
+            return;
+
+        if(ec) {
+            if (ec.value() == boost::system::errc::operation_canceled) {
+                // this websocket has disconnected
+                // TODO delete it from the vector
+                onCloseHandler(socketID_);
+                return;
+            }
+            fail(ec, "read");
+        }
+
+        auto str = beast::buffers_to_string(buffer_.data());
+        onMessageHandler(socketID_, str);
+        buffer_.consume(buffer_.size());
+
+        DoRead();
+    }
+
+    void Send(const std::string& data) {
+        // sync write
+        ws_.write(net::buffer(data));
+    }
+};
+
+void dfws::SendData(Handle hdl, const std::string& data)
+{
+    for (auto& s : sockets) {
+        if (s->socketID_ == hdl) {
+            s->Send(data);
+            return;
+        }
+    }
+    std::cout << "could not find socket with id " << hdl << "\n";
+    exit(1);
+}
+
+void dfwsOnAccept(beast::error_code ec, tcp::socket socket)
+{
+    std::cout << "on accept\n";
+    if(ec)
+        return fail(ec, "accept");
+    // websocket::stream<beast::tcp_stream> ws_(std::move(socket));
+    auto socketPtr = std::make_shared<DfWebsocket>(std::move(socket), lastSocketID++);
+    sockets.push_back(socketPtr);
+    sockets.back()->start();
+
+    // accept another connection
+    acceptor_.async_accept(&dfwsOnAccept);
+}
+
+void dfws::Run(unsigned short port)
+{
+    auto const address = net::ip::make_address("0.0.0.0");
+    beast::error_code ec;
+    tcp::endpoint endpoint{address, port};
+
+    // Open the acceptor
+    acceptor_.open(endpoint.protocol(), ec);
+    if(ec)
```

Comment:
Nit: please ensure that there is a space after each `if`

Reply:
Added spaces after ifs in the whole codebase.